src/custom_component_renderer/book_directory_renderer.rs

```diff
@@ -0,0 +1,141 @@
+use super::dom_manipulator::NodeManipulator;
+use super::{error::RendererError, languages_configuration::LanguagesConfiguration, Component};
+use crate::custom_component_renderer::error::Result;
+use scraper::{Html, Selector};
+use std::io::Write;
+use std::{fs, io::Read, path::Path};
+
+pub(crate) struct BoookDirectoryRenderer {
+    config: LanguagesConfiguration,
+    // Saves on reallocations
+    file_buffer_cache: String,
+    components: Vec<Box<dyn Component>>,
+}
+
+impl BoookDirectoryRenderer {
+    pub(crate) fn new(config: LanguagesConfiguration) -> BoookDirectoryRenderer {
+        BoookDirectoryRenderer {
+            config,
+            file_buffer_cache: String::new(),
+            components: Vec::new(),
+        }
+    }
+
+    pub(crate) fn render_book(&mut self, path: &Path) -> Result<()> {
+        if !path.is_dir() {
+            return Err(RendererError::InvalidPath(format!(
+                "{:?} is not a directory",
+                path
+            )));
+        }
+        self.render_book_directory(path)
+    }
+
+    pub(crate) fn add_component(&mut self, component: Box<dyn Component>) {
+        self.components.push(component);
+    }
+
+    fn render_components(&mut self) -> Result<()> {
+        let mut document = Html::parse_document(&self.file_buffer_cache);
+        for custom_component in &mut self.components {
+            let mut node_ids = Vec::new();
+
+            let selector = Selector::parse(&custom_component.identifier())
+                .map_err(|err| RendererError::InvalidIdentifier(err.to_string()))?;
+            for node in document.select(&selector) {
+                node_ids.push(node.id());
+            }
+            let tree = &mut document.tree;
+            for id in node_ids {
+                let dom_manipulator = NodeManipulator::new(tree, id);
+                custom_component.render(dom_manipulator, &self.config)?;
+            }
+        }
+        self.file_buffer_cache.clear();
+        self.file_buffer_cache = document.html();
+        Ok(())
+    }
+
+    fn process_file(&mut self, path: &Path) -> Result<()> {
+        if path.extension().unwrap_or_default() != "html" {
+            return Ok(());
+        }
+        self.file_buffer_cache.clear();
+        {
+            let mut file = fs::File::open(path)?;
+            file.read_to_string(&mut self.file_buffer_cache)?;
```

Comment:
The read here is used later in `render_components` below? The connection needs to be made explicit, otherwise it can break in a small refactoring.

I would remove the cache and only add it later if you can show a speedup to justify it.